lighthouse-core/computed/metrics/lantern-first-contentful-paint.js

```diff
@@ -24,52 +26,174 @@ class LanternFirstContentfulPaint extends LanternMetric {
+   * @param {(function(CPUNode):boolean)=} extraBlockingCpuNodesToIncludeFilter The function that determines which CPU nodes
+   *    should also be included in our blocking node IDs set.
+   * @return {{definitelyNotRenderBlockingScriptUrls: Set<string>, blockingCpuNodeIds: Set<string>}}
    */
-  static getOptimisticGraph(dependencyGraph, traceOfTab) {
-    const fcp = traceOfTab.timestamps.firstContentfulPaint;
-    const blockingScriptUrls = LanternMetric.getScriptUrls(dependencyGraph, node => {
-      return (
-        node.endTime <= fcp && node.hasRenderBlockingPriority() && node.initiatorType !== 'script'
-      );
-    });
+  static getBlockingCpuData(
+      graph,
+      filterTimestamp,
+      blockingScriptFilter,
+      extraBlockingCpuNodesToIncludeFilter
+  ) {
+    /** @type {Map<string, CPUNode>} */
+    const scriptUrlToNodeMap = new Map();
 
-    return dependencyGraph.cloneWithRelationships(node => {
-      if (node.endTime > fcp && !node.isMainDocument()) return false;
-      // Include EvaluateScript tasks for blocking scripts
+    /** @type {Array<CPUNode>} */
+    const cpuNodes = [];
+    graph.traverse(node => {
       if (node.type === BaseNode.TYPES.CPU) {
-        return node.isEvaluateScriptFor(blockingScriptUrls);
+        // A task is *possibly* render blocking if it *started* before filterTimestamp.
+        // We use startTime here because the paint event can be *inside* the task that was render blocking.
+        if (node.startTime <= filterTimestamp) cpuNodes.push(node);
+
+        // Make sure we still build our script URL to node map.
+        const scriptUrls = node.getEvaluateScriptURLs();
+        for (const url of scriptUrls) {
+          // Use the earliest CPU node we find.
+          const existing = scriptUrlToNodeMap.get(url) || node;
+          scriptUrlToNodeMap.set(url, node.startTime < existing.startTime ? node : existing);
+        }
       }
+    });
+
+    cpuNodes.sort((a, b) => a.startTime - b.startTime);
 
-      // Include non-script-initiated network requests with a render-blocking priority
-      return node.hasRenderBlockingPriority() && node.initiatorType !== 'script';
+    // A script is *possibly* render blocking if it finished loading before filterTimestamp.
+    const possiblyRenderBlockingScriptUrls = LanternMetric.getScriptUrls(graph, node => {
+      return node.endTime <= filterTimestamp && blockingScriptFilter(node);
     });
+
+    // A script is *definitely not* render blocking if its EvaluateScript task started after filterTimestamp.
+    /** @type {Set<string>} */
+    const definitelyNotRenderBlockingScriptUrls = new Set();
+    /** @type {Set<string>} */
+    const blockingCpuNodeIds = new Set();
+    for (const url of possiblyRenderBlockingScriptUrls) {
+      // Lookup the CPU node that had the EvaluateScript for this URL.
+      const cpuNodeForUrl = scriptUrlToNodeMap.get(url);
+
+      // If we can't find it at all, we'll assume it was render-blocking.
```

Comment:
yeah plenty of scripts don't actually get an evaluatescript task, it makes more sense to explain that we can't conclude it's "definitely not render blocking" just because we couldn't find its task, so i've switched to taht :)